### PR TITLE
Lazy instantiation for segmentKillers, segmentMovers, and segmentArchivers

### PR DESCRIPTION
**Pull request by jihoonson**
### Description

Currently, the ingestion task requires for you to specify the AWS region if you have `druid-s3-extensions` loaded _even when your ingestion task doesn't use s3_. This is because `OmniDataSegmentKiller` is being initialized when the Peon starts up which loads all implementations of `DataSegmentKiller` from extensions. This is a bug from UX perspective as the the user should be able to load the s3 extension without AWS region configured if they don't use any s3 stuff. This PR fixes `DataSegmentKiller` as well as  `DataSegmentMover` and `DataSegmentArchiver`, which have the same problem, to accept a client supplier instead of a client, so that the client is initialized when they are actually used. For example, `S3DataSegmentKiller` accepts a supplier of `s3Client` now. Since the AWS region is checked when `s3Client` is initialized and the supplier initializes `s3Client` when the killer is actually used, any ingestion jobs should now be able to run with the s3 extension loaded without AWS region unless they use s3 stuff. The ingestion job will fail when it tries to use extension stuff if something is not configured properly instead of at the beginning of the ingestion job. For example, a kill task that kills segments in s3 will fail when it calls s3 APIs instead of at the beginning of the task if the AWS region is not configured. This change is covered by integration tests as ingestion tests always use the kill task to clean up segments in deep storage after the test is done.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by jihoonson** [comment]
Travis failed because of #12171.

File: indexing-service/src/test/java/org/apache/druid/indexing/common/task/batch/parallel/AbstractParallelIndexSupervisorTaskTest.java [hunk 1]
```diff
@@ -183,8 +183,8 @@
           null
       );
 
-  protected static final double DEFAULT_TRANSIENT_TASK_FAILURE_RATE = 0.3;
-  protected static final double DEFAULT_TRANSIENT_API_FAILURE_RATE = 0.3;
+  protected static final double DEFAULT_TRANSIENT_TASK_FAILURE_RATE = 0.2;
```
Comment (TSFenwick) [inline indexing-service/src/test/java/org/apache/druid/indexing/common/task/batch/parallel/AbstractParallelIndexSupervisorTaskTest.java:186]: Is this a necessary change? seems unrelated to this problem

File: extensions-contrib/aliyun-oss-extensions/src/main/java/org/apache/druid/storage/aliyun/OssStorageDruidModule.java [hunk 2]
```diff
@@ -103,4 +105,11 @@ public OSS initializeOssClient(OssClientConfig inputSourceConfig)
   {
     return inputSourceConfig.buildClient();
   }
+
+  @Provides
+  @LazySingleton
+  public Supplier<OSS> initializeOssClientSupplier(OssClientConfig inputSourceConfig)
+  {
+    return Suppliers.memoize(inputSourceConfig::buildClient);
```
Comment (TSFenwick) [inline extensions-contrib/aliyun-oss-extensions/src/main/java/org/apache/druid/storage/aliyun/OssStorageDruidModule.java:113]: nit: I think leaving a comment here would help people understand that this is lazily supplied to avoid needing to set unnecessary properties

File: extensions-core/azure-extensions/src/main/java/org/apache/druid/storage/azure/AzureStorageDruidModule.java [hunk 3]
```diff
@@ -117,25 +119,30 @@ public void configure(Binder binder)
 
   @Provides
   @LazySingleton
-  public CloudBlobClient getCloudBlobClient(final AzureAccountConfig config)
-      throws URISyntaxException, InvalidKeyException
+  public Supplier<CloudBlobClient> getCloudBlobClient(final AzureAccountConfig config)
```
Comment (TSFenwick) [inline extensions-core/azure-extensions/src/main/java/org/apache/druid/storage/azure/AzureStorageDruidModule.java:122]: like previously mentioned
nit: I think leaving a comment here would help people understand that this is lazily supplied to avoid needing to set unnecessary properties

File: extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsStorageDruidModule.java [hunk 4]
```diff
@@ -81,7 +81,10 @@ public void configure(Binder binder)
              .in(LazySingleton.class);
 
     Binders.dataSegmentPusherBinder(binder).addBinding(SCHEME).to(HdfsDataSegmentPusher.class).in(LazySingleton.class);
-    Binders.dataSegmentKillerBinder(binder).addBinding(SCHEME).to(HdfsDataSegmentKiller.class).in(LazySingleton.class);
+    Binders.dataSegmentKillerBinder(binder)
```
Comment (TSFenwick) [inline extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsStorageDruidModule.java:84]: nit: unnecessary change

**Review by TSFenwick — APPROVED** [review]
looks good to me just some nits

File: indexing-service/src/test/java/org/apache/druid/indexing/common/task/batch/parallel/AbstractParallelIndexSupervisorTaskTest.java (hunk 1 above)
Comment (jihoonson) [reply indexing-service/src/test/java/org/apache/druid/indexing/common/task/batch/parallel/AbstractParallelIndexSupervisorTaskTest.java:186]: Some tests failed because of timeout. As those tests pass with no issue in my computer, one reason that caused the timeout could be too often retries. Reducing failure rate can mitigate too often retry problem.

**Review by clintropolis — APPROVED** [review]

**Comment by jihoonson** [comment]
Travis failed again because of https://github.com/apache/druid/issues/12171.

**Comment by jihoonson** [comment]
Ignoring the flaky test as it doesn't pass even after a couple of retries.
